CRM/Eck/DAO/EckEntityType.php

```diff
@@ -238,63 +203,18 @@ public static function &export($prefix = FALSE) {
    * @return array
    */
   public static function indices($localize = TRUE) {
-    $indices = [];
+    $indices = [
+      'UI_name' => [
+        'name' => 'UI_name',
+        'field' => [
+          0 => 'name',
+        ],
+        'localizable' => FALSE,
+        'unique' => TRUE,
+        'sig' => 'civicrm_eck_entity_type::1::name',
+      ],
+    ];
     return ($localize && !empty($indices)) ? CRM_Core_DAO_AllCoreTables::multilingualize(__CLASS__, $indices) : $indices;
   }
 
-  /**
-   * {@inheritDoc}
-   */
-  public function delete($useWhere = FALSE) {
-    // TODO: Delete entities of this type.
-
-    $result = parent::delete($useWhere);
-
-    // TODO: Delete custom fields in custom groups extending this entity type?
-
-    // Delete custom groups. This has to be done before removing the table due
-    // to FK constraints.
-    foreach (CRM_Eck_BAO_EckEntityType::getCustomGroups($this->name) as $custom_group) {
-      civicrm_api3('CustomGroup', 'delete', ['id' => $custom_group['id']]);
-    }
-
-    // Drop table.
-    $table_name = 'civicrm_eck_' . strtolower($this->name);
-    CRM_Core_DAO::executeQuery(
-      "
-          DROP TABLE IF EXISTS `{$table_name}`;
-          "
-    );
-
-    // Retrieve existing option value for custom-field-extendable object.
-    try {
-      $option_value = civicrm_api3('OptionValue', 'getsingle', [
-        'option_group_id' => 'cg_extend_objects',
-        'value' => 'Eck' . $this->name,
-        'name' => 'civicrm_eck_' . strtolower($this->name),
-      ]);
-      // Delete cg_extend_objects option values.
-      civicrm_api3('OptionValue', 'delete', [
-        'id' => $option_value['id'],
-      ]);
-    }
-    catch (Exception $exception) {
-      // Option value doesn't exist, nothing to do, although this should not
-      // happen.
-    }
-
-    // Delete subtypes.
-    foreach (CRM_Eck_BAO_EckEntityType::getSubTypes($this->name, FALSE) as $sub_type) {
-      civicrm_api3('OptionValue', 'delete', ['id' => $sub_type['id']]);
-    }
-
-    // Flush schema cache.
-    CRM_Core_DAO_AllCoreTables::reinitializeCache();
-
-    // Flush navigation cache.
-    CRM_Core_BAO_Navigation::resetNavigation();
-
-    return $result;
-  }
```

Comment:
This procedure for deleting dependent stuff should then also be moved into a post hook, @colemanw? `ensureEntityType()` does not take care of an entity type being deleted.

Reply:
Ah, I hadn't noticed there was hand-written code in that DAO file which got deleted by regenerating the file.
So, yes you can override the BAO deleteRecord() method and that will work. But I think a neater place to put the code would be in `hook_civicrm_post`. You can do so directly in the BAO file by implementing `HookInterface`. Here's an example: https://github.com/civicrm/civicrm-core/commit/42f0bf162b77ef27b0e93a6b2e882491cdfd8752

Reply:
Interesting, didn't know about that. Now I'm just curious about why that interface is in the `Civi\Test` namespace when it is supposed to identify classes with hook implementations for runtime.

Reply:
Yea, it belongs in a better namespace for sure.